include/ur_api.h

```diff
@@ -1317,6 +1318,46 @@ urDeviceGet(
                                     ///< pNumDevices will be updated with the total number of devices available.
 );
 
+///////////////////////////////////////////////////////////////////////////////
+/// @brief Retrieves devices within a platform selected by ONEAPI_DEVICE_SELECTOR
+///
+/// @details
+///     - Multiple calls to this function will return identical device handles,
+///       in the same order.
+///     - The number and order of handles returned from this function will be
+///       affected by environment variables that filter or select which devices
+///       are exposed through this API.
+///     - A reference is taken for each returned device and must be released
+///       with a subsequent call to ::urDeviceRelease.
+///     - The application may call this function from simultaneous threads, the
+///       implementation must be thread-safe.
+///
+/// @returns
+///     - ::UR_RESULT_SUCCESS
+///     - ::UR_RESULT_ERROR_UNINITIALIZED
+///     - ::UR_RESULT_ERROR_DEVICE_LOST
+///     - ::UR_RESULT_ERROR_ADAPTER_SPECIFIC
+///     - ::UR_RESULT_ERROR_INVALID_NULL_HANDLE
+///         + `NULL == hPlatform`
+///     - ::UR_RESULT_ERROR_INVALID_ENUMERATION
+///         + `::UR_DEVICE_TYPE_VPU < DeviceType`
+///     - ::UR_RESULT_ERROR_INVALID_VALUE
+UR_APIEXPORT ur_result_t UR_APICALL
+urDeviceGetSelected(
```

Comment:
why do we need this in addition to urDeviceGet? Main issue here is that libraries would have duplicated code to discover devices and it would be confusing which one to use. It seems that urDeviceGetSelected is for when ONEAPI_DEVICE_SELECTOR is used, but a library developer doesnt know what the user will use at the end.

I was thinking this logic would be added directly to urDeviceGet, so we have one single entry for device discovery. This is similar to what L0 does, where only zeDeviceGet discovers devices, whether the affinity mask is present or not

Reply:
MPI must always ignore the ONEAPI_DEVICE_SELECTOR, SYCL must never ignore the ONEAPI_DEVICE_SELECTOR, hence "get me a list of all devices" will give two different answers depending on which SW component is asking for the list.

One of the reasons for creating ONEAPI_DEVICE_SELECTOR, despite the previous existence of ZE_AFFINITY_MASK was that Level Zero only has one entry-point that always applies ZE_AFFINITY_MASK but that causes trouble for MPI (and similar SW components).

Reply:
thanks @Wee-Free-Scot . 

Maybe you are referring to the fact that MPI needs to have always visibility of all the devices, while SYCL only of those exposed by ONEAPI_DEVICE_SELECTOR, right?

Reply:
talked with @Wee-Free-Scot.
intention is to have urDeviceGet for full system device visibility (like what MPI needs), and urDeviceGetSelected for filtered device visibility (like what SYCL and OpenMP needs).

In that sense, changes are looking good.